botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+            "service": "s3",
+            "region": "us-east-1",
+            "accountId": "012345678901",
+            "resourceId": ["user", "Development", "product_1234", "*"]
+        }
+        """
+
+        if value is None or not value.startswith("arn:"):
+            return None
+
+        arn_parser = ArnParser()
+        try:
+            arn_dict = arn_parser.parse_arn(value)
+        except InvalidArnException:
+            return None
+
+        # these three components are the only ones that cannot be empty
+        if not all(
+            (arn_dict["partition"], arn_dict["service"], arn_dict["resource"])
+        ):
+            return None
+
+        arn_dict["accountId"] = arn_dict.pop("account")
+
+        resource = arn_dict.pop("resource")
+        delimiter = ":" if ":" in resource else "/"
+        arn_dict["resourceId"] = resource.split(delimiter)
+
+        return arn_dict
+
+    def _string_equals(self, value1, value2):
+        """Evaluates two string values value1 and value2 for equality and
+        returns a boolean if both values match.
+
+        :type value1: str
+        :param value1: a string to compare
+
+        :type value2: str
+        :param value2: a string to compare
+
+        :returns: a bool indicating if value1 and value2 equal each other
+        """
+        if not all(isinstance(val, str) for val in (value1, value2)):
+            raise EndpointInputParametersError(
+                msg=(
+                    f"Both values must be strings.\n"
+                    f"value1: {value1}\n"
+                    f"value2: {value2}"
+                )
+            )
+        return value1 == value2
+
+    def _is_valid_host_label(self, value, allow_subdomains):
+        """Evaluates whether one or more string values are valid host labels per
+        RFC 1123. Each host label must be between 1 and 63 characters, start with
+        a number or letter, and only contain numbers, letters, or hyphens. If
+        allow_subDomains is true, then the provided value may be zero or more dotted
+        subdomains which are each validated per RFC 1123. When allow_subdomains
+        is true, split the inputs on dots, then recurse with allow_subdomains set
+        to false.
+
+        :type value: str
+        :param value: A string that may or may not be a valid RFC 1123 host label
+
+        :type allow_subdomains: bool
+        :param allow_subdomains: Indicating whether the input value can contain
+        zero or more dotted subdomains
+
+        :returns: A bool indicating whether the input value is a valid RFC 1123
+        host label
+        """
+        if value is None or allow_subdomains is False and value.count(".") > 0:
+            return False
+
+        if allow_subdomains is True:
+            return all(
+                self._is_valid_host_label(label, False)
+                for label in value.split(".")
+            )
+
+        return VALID_HOST_LABEL_RE.match(value) is not None
```

Comment:
We currently have two pairs of function and regex that intend to do the same thing:

 * `botocore.utils.check_dns_name` [here](https://github.com/boto/botocore/blob/7d4a3fc41928155a4a1eac04c02b68af64a3a59d/botocore/utils.py#L1267-L1287) and`botocore.utils.LABEL_RE` [here](https://github.com/boto/botocore/blob/7d4a3fc41928155a4a1eac04c02b68af64a3a59d/botocore/utils.py#L95) 
* `botocore.utils.S3ControlEndpointSetter.` [here](https://github.com/boto/botocore/blob/7d4a3fc41928155a4a1eac04c02b68af64a3a59d/botocore/utils.py#L2316-L2317) and `botocore.utils.S3ControlEndpointSetter._HOST_LABEL_REGEX` [here](https://github.com/boto/botocore/blob/7d4a3fc41928155a4a1eac04c02b68af64a3a59d/botocore/utils.py#L2188)

Both of these are less correct/comprehensive than this one here. For example, the first one excludes host labels with uppercase characters.

Could we move the logic and regex from this function to `botocore.utils` instead? That way we can use it to correct the existing functions and consolidate some almost-duplication.

cc @nateprewitt who may have comments on the feasability and utility of fixing these existing functions.

Reply:
That seems fine to me, I agree consolidating is better. If we do change the other customizations, we just need to ensure that we're not materially changing the outcomes of the S3ControlSetter.

Reply:
I'll look into this. Hopefully we don't even need additional unit tests, but I'll make sure.

Reply:
it also looks like `utils.validate_region_name` is doing the same thing. I'll see if I can get rid of that as well.